experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+      codomain::AffineScheme{S,Tc,Uc}, imgs_frac::Vector{Frac{Uc}}
+                           ) where {S,Td,Ud,Tc,Uc}
+    if base_ring(domain) != base_ring(codomain)
+      error( "the base rings of the domain and the codomain do not coincide!" )
+    end
+    k = base_ring(domain)
+    #if domain(pullback) != codomain.R || codomain(pullback) != domain.R
+      #error( "the domain and codomain of the given ring homomorphism is not compatible with the affine schemes." )
+    #end
+    x = new{S,Td,Ud,Tc,Uc}()
+    x.domain = domain
+    x.codomain = codomain
+    x.imgs_frac = imgs_frac
+    # TODO: Implement the checks for the homomorphism to be well defined (on demand).
+    return x
+  end
+end
+
+domain(f::AffSchMorphism) = f.domain
+codomain(f::AffSchMorphism) = f.codomain
+
+# Construct the pullback on the level of coordinate rings 
+# from the fractional representation 
+function pullback(f::AffSchMorphism)
+  if isdefined(f, :pullback)
+    return f.pullback
+  end
+  if !isdefined(f, :imgs_frac )
+    error( "Neither the fractional representation, nor the pullback is defined for this morphism." )
+  end
+  S = ambient_ring(codomain(f))
+  T = ambient_ring(domain(f))
+  R = ambient_ring(root(codomain(f)))
+  # TODO reconstruct the ring homomorphism S -> T from the fractional representation 
+  # and describe it in terms of R -> T with R = ambient_ring(root(codomain(f)))
+
+  n = length( f.imgs_frac )
+  if n != length( gens( R ))
+    error( "Number of variables in the ambient ring of the root does not coincide with the number of images provided for the homomorphism." )
+  end
+  for i in (1:n)
+    p = numerator( f.imgs_frac[i] )
+    q = denominator( f.imgs_frac[i] )
+    u = prod( denoms( domain(f)))
+    (k,a) = coeffs_in_radical( q, u )
+    
+  end
+  #TODO: Finish this once we know how to deal with the radical membership in Oscar.
+
+end
+
+# Construct the fractional representation of the morphism 
+# from the explicit algebra homomorphism. 
+function imgs_frac(f::AffSchMorphism)
+  # first check if this variable is already cached
+  if isdefined(f, Symbol("imgs_frac"))
+    return f.imgs_frac
+  end
+  # if both forms of the morphisms are not to be found, it is not defined at all.
+  if !isdefined( f, :pullback )
+    error( "Neither the fractional representation, nor the pullback is defined for this morphism." )
+  end
+
+  # start reconstructing the fraction representation from 
+  # the explicit ring homomorphism
+  ϕ = pullback(f)
+  # Set up the codomain of a lift of the pullback ϕ to the ambient ring of 
+  # the domain of f. This is a localization of a *free* polynomial ring, 
+  # considered as a subalgebra of the fraction field. 
+  P = ambient_ring(root(domain(f)))
+  F = FractionField(P)
+  den = denoms(domain(f))
+  d = length(den)
+  n = length(gens(P))
+  # prepare a list of the images of the generators 
+  fracs = [ F(x) for x in gens(P) ]
+  fracs = vcat( fracs, [ 1//g for g in den ])
```

Comment:
```suggestion
  append!( fracs, [ 1//g for g in den ])
```
And similar elsewhere